Applications/Spire/Source/Charting/CachedChartModel.cpp

```diff
@@ -1,141 +1,277 @@
-  return load_from_cache({first, last, first, last, limit});
+  if(limit.GetSize() == 0) {
+    return QtPromise([] {
+      return std::vector<Candlestick>();
+    });
+  }
+  auto interval = [&] {
+    if(first == last) {
+      return Interval::closed(first, last);
+    } else {
+      return Interval::open(first, last);
+    }
+  }();
+  auto [bound, cached_interval, queried_interval] = [&] {
+    if(limit.GetType() == SnapshotLimit::Type::HEAD) {
+      return std::make_tuple(first, find_leftmost_cached_overlap(interval),
+        find_leftmost_queried_overlap(interval));
+    } else {
+      return std::make_tuple(last, find_rightmost_cached_overlap(interval),
+        find_rightmost_queried_overlap(interval));
+    }
+  }();
+  if(cached_interval && contains(*cached_interval, bound)) {
+    return on_cached_interval(interval, *cached_interval, limit);
+  } else if(queried_interval && contains(*queried_interval, bound)) {
+    return on_queried_interval(interval, *queried_interval, limit);
+  } else {
+    return on_new_interval(interval, cached_interval, queried_interval,
+      limit);
+  }
 }
 
 connection CachedChartModel::connect_candlestick_slot(
     const CandlestickSignal::slot_type& slot) const {
-  return m_chart_model->connect_candlestick_slot(slot);
+  return connection();
 }
 
-QtPromise<std::vector<Candlestick>> CachedChartModel::load_from_cache(
-    const LoadInfo& info) {
-  return m_cache.load(info.m_requested_first, info.m_requested_last,
-      info.m_limit).then(
-    [=] (auto result) {
-      if(static_cast<int>(result.Get().size()) == info.m_limit.GetSize() ||
-          contains(m_ranges, continuous_interval<Scalar>::closed(
-          info.m_requested_first, info.m_requested_last))) {
-        return QtPromise(std::move(result.Get()));
-      }
-      auto load_first = info.m_requested_first;
-      auto load_last = info.m_requested_last;
-      if(info.m_limit.GetType() == SnapshotLimit::Type::HEAD) {
-        for(auto& range : m_ranges) {
-          auto first = continuous_interval<Scalar>::closed(load_first,
-            load_first);
-          if(intersects(first, range)) {
-            load_first = range.upper();
-          }
-          if(exclusive_less(first, range)) {
-            load_last = std::min(load_last, range.lower());
-            break;
-          }
-        }
+QtPromise<std::vector<Candlestick>> CachedChartModel::on_cached_interval(
+    const Interval& interval, const Interval& cached_interval,
+    const SnapshotLimit& limit) {
+  auto is_head = limit.GetType() == SnapshotLimit::Type::HEAD;
+  auto [start, end] = [&] {
+    if(is_head) {
+      return std::make_pair(interval.lower(),
+        std::min(interval.upper(), cached_interval.upper()));
+    } else {
+      return std::make_pair(std::max(interval.lower(),
+        cached_interval.lower()), interval.upper());
+    }
+  }();
+  return m_cache.load(start, end, limit).then(
+    [=, start = start, end = end] (auto result) {
```

Comment:
You can remove the ```start = start```, ```end = end```.

Reply:
I did it to work around the intellisense bug for capturing structured bindings. However, I've just confirmed it and MSBuild seems to run fine without it, just the ugly intellisense errors when running VS.

Reply:
Intellisense sucks, doesn't even use the same compiler front end as MSVC.